File: src/main/java/com/official/memento/todo/service/ToDoPrioritizationService.java
package com.official.memento.todo.service;

import com.official.memento.global.exception.ErrorCode;
import com.official.memento.global.exception.InvalidAiRequestException;
import com.official.memento.member.service.result.MemberPersonalInfoResult;
import com.official.memento.member.service.usecase.MemberPersonalInfoGetUseCase;
import com.official.memento.orderinfo.domain.OrderInfo;
import com.official.memento.orderinfo.service.usecase.OrderInfoGetUseCase;
import com.official.memento.orderinfo.service.usecase.OrderInfoUpdateUseCase;
import com.official.memento.tag.service.result.TagResult;
import com.official.memento.tag.service.usecase.TagGetUseCase;
import com.official.memento.todo.domain.entity.ToDo;
import com.official.memento.todo.domain.repository.ToDoRepository;
import com.official.memento.todo.domain.vo.ClaudeAiChatClientOutputPort;
import com.official.memento.todo.service.command.ToDoPrioritizationCommand;
import com.official.memento.todo.service.usecase.ToDoPrioritizationUseCase;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
public class ToDoPrioritizationService implements ToDoPrioritizationUseCase {

    private final ClaudeAiChatClientOutputPort toDoAiChatClientOutputPort;
    private final ToDoRepository toDoRepository;
    private final OrderInfoGetUseCase orderInfoGetUseCase;
    private final OrderInfoUpdateUseCase orderInfoUpdateUseCase;
    private final TagGetUseCase tagGetUseCase;
    private final MemberPersonalInfoGetUseCase memberPersonalInfoGetUseCase;

    @Override
    @Transactional
    public List<List<ToDo>> prioritizeWeekly(ToDoPrioritizationCommand command) {
        MemberPersonalInfoResult memberPersonalInfo = memberPersonalInfoGetUseCase.findByMemberId(command.memberId());
        List<List<ToDo>> prioritizedToDoList = new ArrayList<>();

        LocalDate currentDate = command.targetDate();

        for (int i = 0; i <= 6; i++) {
            List<ToDo> toDoList = toDoRepository.findAllByMemberIdAndStartDate(
                    command.memberId(),
                    currentDate
            );

            final LocalDate finalCurrentDate = currentDate;
            List<Double> orderList = toDoList.stream()
                    .map(toDo -> orderInfoGetUseCase.findByToDoIdAndDate(toDo.getId(), finalCurrentDate).getOrderNum())
                    .toList();

            List<ToDo> prioritizedToDos = toDoAiChatClientOutputPort.prioritizeTodo(
                    toDoList,
                    orderList,
                    memberPersonalInfo.toPersonalInfoString()
            ).stream()
                    .map(prioritizedToDo -> {
                        OrderInfo orderInfo = orderInfoGetUseCase.findByToDoIdAndDate(
                                prioritizedToDo.id(),
                                finalCurrentDate
                        );
                        orderInfoUpdateUseCase.updateOrderNum(orderInfo, prioritizedToDo.order());

                        ToDo toDo = toDoRepository.findById(prioritizedToDo.id());
                        TagResult tag = tagGetUseCase.findById(toDo.getTagId());

                        return toDo.toBuilder()
                                .priorityUrgency((double) prioritizedToDo.urgency())
                                .priorityImportance((double) prioritizedToDo.importance())
                                .priorityValue((double) prioritizedToDo.priority())
                                .orderNum(prioritizedToDo.order())
                                .tagName(tag.name())
                                .tagColor(tag.color())
                                .build();
                    })
                    .toList();

            prioritizedToDoList.add(i, prioritizedToDos);
            currentDate = currentDate.plusDays(1);
        }

        return prioritizedToDoList;
    }

    @Override
    @Transactional
    public List<ToDo> prioritizeDaily(ToDoPrioritizationCommand command) {
        List<ToDo> toDoList = toDoRepository.findAllByMemberIdAndStartDate(
                command.memberId(),
                command.targetDate()
        );

        if (toDoList.isEmpty()) {
            throw new InvalidAiRequestException(ErrorCode.INVALID_AI_PRIORITIZATION_REQUEST);
        }

        MemberPersonalInfoResult memberPersonalInfo = memberPersonalInfoGetUseCase.findByMemberId(command.memberId());

        List<Double> orderList = toDoList.stream()
                .map(toDo -> orderInfoGetUseCase.findByToDoIdAndDate(toDo.getId(), command.targetDate()).getOrderNum())
                .toList();

        return toDoAiChatClientOutputPort.prioritizeTodo(
                toDoList,
                orderList,
                memberPersonalInfo.toPersonalInfoString()
        ).stream()
                .map(prioritizedToDo -> {
                    OrderInfo orderInfo = orderInfoGetUseCase.findByToDoIdAndDate(
                            prioritizedToDo.id(),
                            command.targetDate()
                    );
                    orderInfoUpdateUseCase.updateOrderNum(orderInfo, prioritizedToDo.order());

                    ToDo toDo = toDoRepository.findById(prioritizedToDo.id());
                    TagResult tag = tagGetUseCase.findById(toDo.getTagId());

                    return toDo.toBuilder()
                            .priorityUrgency((double) prioritizedToDo.urgency())
                            .priorityImportance((double) prioritizedToDo.importance())
                            .priorityValue((double) prioritizedToDo.priority())
                            .orderNum(prioritizedToDo.order())
                            .tagName(tag.name())
                            .tagColor(tag.color())
                            .build();
                })
                .toList();
    }
}

File: src/main/kotlin/com/official/memento/todo/service/ToDoPrioritizationService.kt
package com.official.memento.todo.service

import com.official.memento.orderinfo.domain.OrderInfoRepository
import com.official.memento.tag.domain.TagRepository
import com.official.memento.todo.domain.ToDo
import com.official.memento.todo.domain.ToDoRepository
import com.official.memento.todo.domain.ToDoTagRepository
import com.official.memento.todo.domain.vo.ClaudeAiChatClientOutputPort
import com.official.memento.todo.service.command.ToDoPrioritizationCommand
import org.springframework.stereotype.Service

@Service
class ToDoPrioritizationService(
    private val toDoAiChatClientOutputPort: ClaudeAiChatClientOutputPort,
    private val toDoRepository: ToDoRepository,
    private val orderInfoRepository: OrderInfoRepository,
    private val toDoTagRepository: ToDoTagRepository,
    private val tagRepository: TagRepository,
) : ToDoPrioritizationUseCase {
    override fun prioritizeWeekly(command: ToDoPrioritizationCommand) : List<List<ToDo>> {
        val prioritizedToDoList = mutableListOf<List<ToDo>>()
        for (i in 0..6) {
            val toDoList = toDoRepository.findAllByMemberIdAndStartDate(command.memberId, command.targetDate)
            val orderList =
                toDoList.map {
                    orderInfoRepository.findByToDoIdAndDate(it.id, command.targetDate).orderNum
                }.toList()
            prioritizedToDoList.add(
                i,
                toDoAiChatClientOutputPort.prioritizeTodo(toDoList, orderList).map {
                    val orderInfo = orderInfoRepository.findByToDoIdAndDate(it.id, command.targetDate)
                    orderInfoRepository.updateOrderNum(orderInfo, it.order)
                    val toDo = toDoRepository.findById(it.id)
                    val toDoTag = toDoTagRepository.findByToDoId(it.id)
                    val tag = tagRepository.findById(toDoTag.tagId)
                    ToDo.withIdAndTagAndOrder(
                        toDo.id,
                        toDo.memberId,
                        toDo.groupId,
                        toDo.startDate,
                        toDo.description,
                        toDo.endDate,
                        toDo.isCompleted,
                        toDo.repeatOption,
                        toDo.repeatExpiredDate,
                        it.urgency.toDouble(),
                        it.importance.toDouble(),
                        it.priority.toDouble(),
                        toDo.priorityType,
                        toDo.type,
                        toDo.createdAt,
                        toDo.updatedAt,
                        it.order,
                        tag.id,
                        tag.name,
                        tag.color,
                    )
                }.toList(),
            )
            command.targetDate.plusDays(1)
        }
        return prioritizedToDoList
    }

    override fun prioritizeDaily(command: ToDoPrioritizationCommand): List<ToDo> {
        val toDoList = toDoRepository.findAllByMemberIdAndStartDate(command.memberId, command.targetDate)
        val orderList =
            toDoList.map {
                orderInfoRepository.findByToDoIdAndDate(it.id, command.targetDate).orderNum
            }.toList()
        return toDoAiChatClientOutputPort.prioritizeTodo(toDoList, orderList).map {
                val orderInfo = orderInfoRepository.findByToDoIdAndDate(it.id, command.targetDate)
                orderInfoRepository.updateOrderNum(orderInfo, it.order)
                val toDo = toDoRepository.findById(it.id)
                val toDoTag = toDoTagRepository.findByToDoId(it.id)
                val tag = tagRepository.findById(toDoTag.tagId)
                ToDo.withIdAndTagAndOrder(
                    toDo.id,
                    toDo.memberId,
                    toDo.groupId,
                    toDo.startDate,
                    toDo.description,
                    toDo.endDate,
                    toDo.isCompleted,
                    toDo.repeatOption,
                    toDo.repeatExpiredDate,
                    it.urgency.toDouble(),
                    it.importance.toDouble(),
                    it.priority.toDouble(),
                    toDo.priorityType,
                    toDo.type,
                    toDo.createdAt,
                    toDo.updatedAt,
                    it.order,
                    tag.id,
                    tag.name,
                    tag.color,
                )
            }.toList()
    }

}
